### PR TITLE
hardens the aggregated API e2e tests in an HA setup

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
This PR makes the aggregated e2e tests more stable in an HA environment by checking if the sample server was unregistered from all deployed Kube API servers **before** tearing down the deployment. This ensures that the whole process is **graceful** and it is safe to remove the binary from the platform.

It also provides a helper as a general-purpose method of ensuring `x` is true on all deployed API servers. 
For now, only one way of detecting all deployed API servers is supported. This could be extended in the future.
Running the helper on an unsupported platform only prints a warning message in the logs.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by liggitt** [comment]
if I'm understanding correctly, this does the following:

1. looks up kubernetes service endpoints once
2. runs a pod proxying ports to those endpoints
3. runs `kubectl port-forward` on the machine running the e2e's to communicate with that pod
4. connects to `localhost:<port>` in a loop to check results from various API servers

Maybe it's just me, but I find that hard to follow, and it seems like it would be fragile:
* How does it handle the kubernetes service endpoints changing?
* How does the proxy pod get cleaned up?
* I don't think we want to add `kubectl port-forward` invocations as part of the e2e run

If the goal is to wait until the content of a given URL has converged for all Kubernetes API servers, it seems like that would be a better building block to have in agnhost. An invocation that would fetch a given path from all endpoints servicing the Kubernetes service, waiting until the responses converged and a given status code was seen, or specific content was seen, or specific content was not seen.

```go
// CoherentOptions indicates the checks to run on the results once the same result is received from all Kubernetes service endpoints
type CoherentResponseOptions struct {
  // wait until this status is received from all
  Status int
  // wait until the coherent response contains all the strings in this list
  Contains []string
  // wait until the coherent response does not contain any of the strings in this list
  DoesNotContain []string
}

func WaitForCoherentResponse(ctx context.Contetx, path string, opts CoherentResponseOptions) {...}
```

which could then be used like this:

```go
f.ExpectNoError(WaitForCoherentResponse(ctx, "/apis/wardle.example.com", CoherentResponseOptions{Status: 404}))
```

**Comment by smarterclayton** [comment]
>  the goal is to wait until the content of a given URL has converged for all Kubernetes API servers, it seems like that would be a better building block to have in agnhost

so for authenticated urls you have to give that permission to the service account?  i guess i could buy having agnhost have a kube-api specific watch function... trying to think about external facing implications (whether someone could want to have a conformant cluster that exposes different things downwards and upwards)

**Comment by liggitt** [comment]
the current uses for this I can envision in an e2e test would be around discovery / openapi endpoints... those are expected to be available to authenticated users already, and are part of conformance, so I'd expect them to be consistent

the kubectl / portforwarding / point-in-time proxy pod approach just seems fragile to me

File: test/e2e/framework/apihelper/helper.go [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+				return ok, err
+			}
+		}
+		return true, nil
+	}
+}
+
+// setupAPIServersProxyPodAndPortForward a convenience method that creates and runs a pod that proxies connections to the API servers.
+// It also uses kubectl port-forward to route local connections to that pod.
+func setupAPIServersProxyPodAndPortForward(client clientset.Interface, namespace string) ([]int, func(), error) {
```
Comment (aojea) [inline test/e2e/framework/apihelper/helper.go:136]: we have a nice dialer that does this, it was added recently https://github.com/kubernetes/kubernetes/pull/102050/commits/5e9076da93c8bd8d51dd9c6d2d4f87d5a269e586#diff-1603c90cc2b55449a34718b2333960968302a0e5d8c9d43cafeca2cae8c0f205
